test/infrastructure/docker/controllers/hollowmachine_controller.go

```diff
@@ -0,0 +1,221 @@
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// HollowMachineReconciler reconciles a HollowMachine object.
+type HollowMachineReconciler struct {
+	client.Client
+	Tracker *remote.ClusterCacheTracker
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=hollowmachines,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=hollowmachines/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;machines,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=secrets;,verbs=get;list;watch
+
+// Reconcile handles HollowMachine events.
+func (r *HollowMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the HollowMachine instance.
+	hollowMachine := &infrav1.HollowMachine{}
+	if err := r.Client.Get(ctx, req.NamespacedName, hollowMachine); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the HollowKubelet.
+	machine, err := util.GetOwnerMachine(ctx, r.Client, hollowMachine.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if machine == nil {
+		log.V(2).Info("Waiting for Machine Controller to set OwnerRef on the HollowMachine")
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("machine", machine.Name)
+
+	if util.IsControlPlaneMachine(machine) {
+		return ctrl.Result{}, errors.New("HollowMachine can't be used for control plane machines")
+	}
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		log.Info("HollowMachine owner HollowKubelet is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info(fmt.Sprintf("Please associate this machine with a cluster using the label %s: <name of cluster>", clusterv1.ClusterLabelName))
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	// Return early if the object or Cluster is paused.
+	if annotations.IsPaused(cluster, hollowMachine) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(hollowMachine, r)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Always attempt to Patch the HollowMachine object and status after each reconciliation.
+	defer func() {
+		// TODO: investigating patching errors on delete
```

Comment:
Personally I find the defer/patch/rerr pattern more difficult to follow than being explicit.
Why `if rerr == nil {`? `rerr` seems to be always nil.

Reply:
defer is the standard approach in CAPI controllers.
`rerr` value depends on the return statements down in the same function